arch/x86/core/userspace.S

```diff
@@ -8,46 +8,74 @@
 #include <arch/x86/asm.h>
 #include <arch/cpu.h>
 #include <offsets_short.h>
+#include <syscall.h>
 
 /* Exports */
 GTEXT(_x86_syscall_entry_stub)
 GTEXT(_x86_userspace_enter)
 
 /* Imports */
-GTEXT(_k_syscall_entry)
+GTEXT(_k_syscall_table)
 
 /* Landing site for syscall SW IRQ. Marshal arguments and call C function for
- * further processing.
+ * further processing. We're on the kernel stack for the invoking thread.
  */
 SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
-	push	%esi		/* call_id */
+	sti			/* re-enable interrupts */
+
+	/* Stash these values so we don't clobber them */
+	push	%ebp
+	push	%esi
+
+	/* call_id is in ESI. bounds-check it, must be less than
+	 * K_SYSCALL_LIMIT
+	 */
+	cmp	$K_SYSCALL_LIMIT, %esi
+	jae	_bad_syscall
+
+_id_ok:
+	/* replace the call ID in ESI with the actual function pointer to call
+	 * by looking it up in the system call dispatch table
+	 */
+	xor	%ebp, %ebp
+	mov	_k_syscall_table(%ebp, %esi, 4), %esi
+
+	/* Marshal arguments per calling convention to match what is expected
+	 * for _k_syscall_handler_t functions
+	 */
+	push	%esp		/* syscall stack frame argument */
 	push	%edi		/* arg5 */
 	push	%ebx		/* arg4 */
-#ifndef CONFIG_X86_IAMCU
+
+	/* For IAMCU we don't need to push these three for the calling
+	 * convention but we need to make sure on return that these were the
+	 * same values when we got here as the compiler has no idea that the
+	 * 'int' instruction at the call site could have side effects.
+	 */
```

Comment:
On retrospect I think this is wrong. I think I need to adjust the system call invocations to make some registers output operands